### ISSUE TITLE
Strange promises behaviour in v3.16.x

### ISSUE THREAD

**Issue by DRoet** [fastify/fastify#3107]
## 💥 Regression Report

Sorry for the long reproduction, took me a long time to dumb it down and I don't fully understand what is happening under the hood.

for some reason Fastify started throwing reply-errors after upgrading to 3.16.2:
`FastifyError: Promise may not be fulfilled with 'undefined' when statusCode is not 204`

When I change the `reply.code(200).send('OK')` to `await reply.code(200).send('OK')` in my example the error dissapears...

Whenever I remove the fastify-session + redis middleware it seems fine too.

## Last working version

Worked up to version: 3.15.1

Stopped working in version: 3.16.2

## To Reproduce

```js
const fastifyCookie = require('fastify-cookie')
const fastifySession = require('fastify-session')
const RedisStore = require('connect-redis')(fastifySession)
const redis = require('redis')
const fastify = require('fastify')({
   logger: true,
})

const redisClient = () => {
   const client = redis.createClient({
      host: 'localhost',
      port: 6379,
   })
   client.on('connect', function () {
      console.log('connected!')
   })
   return client
}

fastify.register(fastifyCookie)
fastify.register(fastifySession, {
   store: new RedisStore({ client: redisClient() }),
   secret: '12345678901234567890123456789012345678901234567890',
   saveUninitialized: true,
   cookie: {
      secure: false,
   },
})

fastify.get('/', async (request, reply) => {
   await new Promise(resolve => resolve())

   // this is OK
   // await reply.code(200).send('OK')
   
   // this throws the error
   reply.code(200).send('OK')
})

const start = async () => {
   try {
      await fastify.listen(5000)
   } catch (err) {
      fastify.log.error(err)
      process.exit(1)
   }
}
start()
```

## Expected behavior

```js
3.15.1:

{"level":30,"time":1622202218911,"pid":12964,"hostname":"DR","msg":"Server listening at http://127.0.0.1:5000"}
connected!
{"level":30,"time":1622202222116,"pid":12964,"hostname":"DR","reqId":"req-1","req":{"method":"GET","url":"/","hostname":"localhost:5000","remoteAddress":"127.0.0.1","remotePort":63196},"msg":"incoming request"}
{"level":30,"time":1622202222124,"pid":12964,"hostname":"DR","reqId":"req-1","res":{"statusCode":200},"responseTime":7.853700011968613,"msg":"request completed"}

3.16.2:

{"level":30,"time":1622202255749,"pid":15216,"hostname":"DR","msg":"Server listening at http://127.0.0.1:5000"}
connected!
{"level":30,"time":1622202259005,"pid":15216,"hostname":"DR","reqId":"req-1","req":{"method":"GET","url":"/","hostname":"localhost:5000","remoteAddress":"127.0.0.1","remotePort":63212},"msg":"incoming request"}
{"level":50,"time":1622202259009,"pid":15216,"hostname":"DR","reqId":"req-1","err":{"type":"FastifyError","message":"Promise may not be fulfilled with 'undefined' when statusCode is not 204","stack":"FastifyError: Promise may not be fulfilled with 'undefined' when statusCode is not 204\n    at fastify\\lib\\wrapThenable.js:28:30\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)","name":"FastifyError","code":"FST_ERR_PROMISE_NOT_FULFILLED","statusCode":500},"msg":"Promise may not be fulfilled with 'undefined' when statusCode is not 204"}
{"level":30,"time":1622202259013,"pid":15216,"hostname":"DR","reqId":"req-1","res":{"statusCode":200},"responseTime":7.443599998950958,"msg":"request completed"}
```

## Your Environment

- *node version*: 16
- *fastify version*: >=3.16.0
- *os*:  Windows / Linux
- *any other relevant information*

**Comment by DRoet** [comment]
Maybe I misunderstand. am I required to **always** add return/await to `reply.send()` inside of async functions?

**Comment by jsumners** [comment]
I think this is solved by https://github.com/fastify/fastify/pull/2702 in `fastify@4` when it releases.

**Comment by mattbishop** [comment]
I am encountering this error with 3.16 as well, but in my case I am sending a Stream:

```js
    // last line of my async handler function
    reply
      .type(CONTENT_TYPES.ND_JSON)
      .send(replyStream);
}
```
I debugged through the code and the problem seems to be wrapThenable's payload value is undefined, even though it is the correct type in reply.js' send() function.

With 3.15 and earlier, I never had to set the `reply.sent` property.

**Comment by mcollina** [comment]
The revert shipped in v3.17.0, the new behavior will land in v4.0.0.

@mattbishop FWIW in your code you had to add a `return reply` at the end to fix, however we should not have broken your code in the v3 series.

**Comment by zekth** [comment]
@mcollina might be worth to write (at least try) a migration helper like we did for `tap 14` to `tap 15`, to find the `send()` instructions without `return` statements.

**Comment by mcollina** [comment]
I think that is going to be harder :(, as it would require some AST parsing.

**Comment by mattbishop** [comment]
@mcollina maybe It would be easier to return the Stream instead of the reply? I agree with the concept of not returning undefined Promises, and it shouldn't be too much effort to support returning a Stream to indicate a successful request this way.

**Comment by mcollina** [comment]
Returning the stream already works as expected.

**Comment by mattbishop** [comment]
I didn’t know I could return the stream, that works. Thanks!

—
Matt Bishop

**Comment by mcollina** [comment]
reply.send() effectively forks the promise chain.
